modules/rabbitmq/src/main/java/org/apache/axis2/transport/rabbitmq/RabbitMQSender.java

```diff
@@ -58,81 +59,161 @@ public void init(ConfigurationContext cfgCtx, TransportOutDescription transportO
+                if (StringUtils.isNotEmpty(queueName) && StringUtils.isEmpty(routingKey)) {
+                    routingKey = queueName; // support the backward compatibility
+                } else if (StringUtils.isEmpty(queueName) && StringUtils.isEmpty(routingKey)) {
+                    routingKey = targetEPR.substring(targetEPR.indexOf("/") + 1, targetEPR.indexOf("?"));
+                }
+
+                // send the message
+                senderType = getSenderType(msgCtx, epProperties);
+                if (senderType == SenderType.PUBLISHER_CONFIRMS) {
+                    channel = rabbitMQConfirmChannelPool.borrowObject(factoryName);
+                } else {
+                    channel = rabbitMQChannelPool.borrowObject(factoryName);
+                }
+                RabbitMQMessageSender sender = new RabbitMQMessageSender(channel, factoryName, senderType);
+                response = sender.send(routingKey, msgCtx, epProperties);
+            } catch (Exception e) {
+                log.error("Error occurred while sending message out.", e);
+                returnToPool(factoryName, channel, senderType);
+                channel = null;
+            } finally {
+                returnToPool(factoryName, channel, senderType);
+            }
+
+            // inject message to the axis engine if a response received
+            if (response != null) {
+                MessageContext responseMsgCtx = createResponseMessageContext(msgCtx);
+                String contentType = RabbitMQUtils.buildMessage(
+                        response.getProperties(), response.getBody(), responseMsgCtx);
+                handleIncomingMessage(responseMsgCtx, RabbitMQUtils.getTransportHeaders(response.getProperties()),
+                        RabbitMQUtils.getSoapAction(response.getProperties()),
+                        contentType);
+            }
+        } else if (outTransportInfo instanceof RabbitMQOutTransportInfo) {
+            // execute when publishing a message to the replyTo queue in request-response flow
+            RabbitMQOutTransportInfo transportOutInfo = (RabbitMQOutTransportInfo) outTransportInfo;
+            String factoryName = transportOutInfo.getConnectionFactoryName();
+            Channel channel = null;
+            try {
+                // get the correlationId and replyTo queue from the transport headers
+                Map<String, Object> transportHeaders =
+                        (Map<String, Object>) msgCtx.getProperty(msgCtx.TRANSPORT_HEADERS);
+                msgCtx.setProperty(
+                        RabbitMQConstants.CORRELATION_ID, transportHeaders.get(RabbitMQConstants.CORRELATION_ID));
+                String replyTo = (String) transportHeaders.get(RabbitMQConstants.RABBITMQ_REPLY_TO);
+
+                // set rabbitmq properties
+                Hashtable<String, String> rabbitMQProperties = new Hashtable<>();
```

Comment:
Shall we use HashMap? We don't need a synchronized data structure right? I assume we don't access this from different threads right?
```suggestion
                Map<String, String> rabbitMQProperties = new HashMap<>();
```

Reply:
It's difficult to change here because the RabbitMQMessageSender#send method requires a Hashtable object as the parameter. When `if (targetEPR != null) ` situation we are using `BaseUtils.getEPRProperties(targetEPR)` to get the parameters which returns a Hashtable.

Reply:
The above suggestion accommodate by replacing `Map<String, String>`.